src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1250,34 +1317,103 @@ public void OnStartLine(HttpMethod method, HttpVersion version, Span<byte> targe
                 }
             }
 
-            var normalizedTarget = PathNormalizer.RemoveDotSegments(requestUrlPath);
-            if (method != HttpMethod.Custom)
+            QueryString = query.GetAsciiStringNonNullCharacters();
+            RawTarget = rawTarget;
+            SetNormalizedPath(requestUrlPath);
+        }
+
+
+        private void OnAuthorityFormTarget(HttpMethod method, Span<byte> target, Span<byte> line)
+        {
+            // TODO Validate that target is a correct host[:port] string.
+            // Reject as 400 if not. This is just a quick scan for invalid characters
+            // but doesn't check that the target fully matches the URI spec.
+            for (var i = 0; i < target.Length; i++)
+            {
+                var ch = (char)target[i];
+                if (!char.IsLetterOrDigit(ch) && ch != ':' && ch != '.')
+                {
+                    if (Log.IsEnabled(LogLevel.Information))
```

Comment:
That's the extra work to avoid 😉  push it to right of throw and combine the two
```csharp
private BadHttpRequestException GetRequestLineRejection(Span<byte> line)
{
    const int MaxRequestLineError = 32;
    if (Log.IsEnabled(LogLevel.Information))
    {
         var line = line.GetAsciiStringEscaped(MaxRequestLineError);
         return BadHttpRequestException.GetException(RequestRejectionReason.InvalidRequestLine, line);
    }
    else
    {
         return BadHttpRequestException.GetException(RequestRejectionReason.InvalidRequestLine);
    }
}

private void ThrowRequestLineRejection(Span<byte> line)
{
    throw GetRequestLineRejection(line)
}
```
Then
```csharp
if (!char.IsLetterOrDigit(ch) && ch != ':' && ch != '.')
{
    ThrowRequestLineRejection(line);
}
```
Keep the main body tight with only code that's mostly always executed; push everything else out